### PR TITLE
Add support for AGP 7.0.0-alpha01

### PR DESCRIPTION
**Pull request by ghale**
This adds support for AGP 7.0.0-alpha01 which requires JDK11 to run.  Unfortunately, TestKit does not provide a way to run with different JDKs, so to handle this, this PR adds two things:

- A `testWithJDK11` test task which runs tests using JDK11.  The original `test` task runs the tests for AGP versions that need JDK8 and `testWithJDK11` runs the tests for AGP versions that need JDK11 (currently only 7.0.0-alpha01).
- A `MultiVersionTest` junit category which allows us to differentiate between tests that only need to run once and tests that need to be run with both test tasks.

### CODE REVIEW

File: src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy [hunk 1]
```diff
@@ -99,10 +95,9 @@ class AndroidCacheFixPlugin implements Plugin<Project> {
     void apply(Project project) {
         workarounds.addAll(initializeWorkarounds(project))
 
-
         if (!isSupportedAndroidVersion(project)) {
             if (isMaybeSupportedAndroidVersion(project)) {
-                project.logger.warn("WARNING: Android plugin ${CURRENT_ANDROID_VERSION} has not been tested with this version of the Android cache fix plugin, although it may work.  We test against only the latest patch release versions of Android Gradle plugin: ${SUPPORTED_ANDROID_VERSIONS.join(", ")}.  If ${CURRENT_ANDROID_VERSION} is newly released, we may not have had a chance to release a version tested against it yet.  Proceed with caution.  You can suppress this warning with with -D${IGNORE_VERSION_CHECK_PROPERTY}=true.")
+                Warnings.MAYBE_SUPPORTED_ANDROID_VERSION.warnOnce(project.logger)
```
Comment (runningcode) [inline src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy:100]: Nice! Warn once!

File: src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy [hunk 2]
```diff
@@ -162,270 +157,20 @@ class AndroidCacheFixPlugin implements Plugin<Project> {
...
-                }
+        void warnOnce(org.gradle.api.logging.Logger logger) {
+            if (!warned.getAndSet(true)) {
+                logger.warn(warning)
             }
         }
     }
-     */
-
-    /**
```
Comment (runningcode) [inline src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy:238]: good idea to delete this code! thanks.

File: src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy [hunk 3]
```diff
@@ -502,10 +520,12 @@ class CrossVersionOutcomeAndRelocationTest extends AbstractTest {
...
         builder.expect(':app:writeReleaseAppMetadata', FROM_CACHE)
-        // New non-cacheable tasks in 4.2.0-alpha10:
-        builder.expect(':app:writeReleaseApplicationId', SUCCESS)
-        builder.expect(':app:analyticsRecordingRelease', SUCCESS)
         builder.expect(':app:writeDebugSigningConfigVersions', FROM_CACHE)
         builder.expect(':app:writeReleaseSigningConfigVersions', FROM_CACHE)
     }
+
+    static void android70xOrHIgherExpectations(ExpectedOutcomeBuilder builder) {
+        builder.expect(':app:desugarDebugFileDependencies', FROM_CACHE)
```
Comment (runningcode) [inline src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy:528]: oy, and now its cacheable. I hope they fixed the overlapping outputs issue!

File: src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy [hunk 4]
```diff
@@ -162,270 +157,20 @@ class AndroidCacheFixPlugin implements Plugin<Project> {
...
-        }
-    }
-     */
+    private enum Warnings {
+        MAYBE_SUPPORTED_ANDROID_VERSION("WARNING: Android plugin ${CURRENT_ANDROID_VERSION} has not been tested with this version of the Android cache fix plugin, although it may work.  We test against only the latest patch release versions of Android Gradle plugin: ${SUPPORTED_ANDROID_VERSIONS.join(", ")}.  If ${CURRENT_ANDROID_VERSION} is newly released, we may not have had a chance to release a version tested against it yet.  Proceed with caution.  You can suppress this warning with with -D${IGNORE_VERSION_CHECK_PROPERTY}=true.")
 
-    /**
-     * Filter the Java annotation processor output folder from compiler arguments to avoid absolute path.
+        private final String warning
+        private final AtomicBoolean warned = new AtomicBoolean()
```
Comment (runningcode) [inline src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy:164]: This is cool! For some reason, I thought that the Plugin object would be created once per project where it is applied.

File: src/test/groovy/org/gradle/android/MultiVersionTest.groovy [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package org.gradle.android
+
+/**
+ * Represents tests that should be run with
```
Comment (runningcode) [inline src/test/groovy/org/gradle/android/MultiVersionTest.groovy:4]: `run with...`? :)

File: src/test/groovy/org/gradle/android/WorkaroundTest.groovy [hunk 6]
```diff
@@ -1,12 +1,17 @@
 package org.gradle.android
 
+import org.gradle.api.JavaVersion
+import org.junit.Assume
+import org.junit.experimental.categories.Category
 import spock.lang.Specification
 import spock.lang.Unroll
 
+@Category(MultiVersionTest)
```
Comment (runningcode) [inline src/test/groovy/org/gradle/android/WorkaroundTest.groovy:9]: Is this a `MultiVersionTest` if it is only run against java8? Or does multiversion mean multiple android versions?

File: src/test/groovy/org/gradle/android/PluginApplicationTest.groovy [hunk 7]
```diff
@@ -35,10 +35,12 @@ class PluginApplicationTest extends AbstractTest {
...
             .withArguments("tasks", "--stacktrace")
             .build()
-        result.output =~ /WARNING: Android plugin ${quote("3.6.1")} has not been tested with this version of the Android cache fix plugin, although it may work.  We test against only the latest patch release versions of Android Gradle plugin: ${Versions.SUPPORTED_ANDROID_VERSIONS.join(", ")}.  If 3.6.1 is newly released, we may not have had a chance to release a version tested against it yet.  Proceed with caution.  You can suppress this warning with with -Dorg.gradle.android.cache-fix.ignoreVersionCheck=true./
+        result.output.readLines().findAll {
+            it =~ /WARNING: Android plugin ${quote("3.6.1")} has not been tested with this version of the Android cache fix plugin, although it may work.  We test against only the latest patch release versions of Android Gradle plugin: ${Versions.SUPPORTED_ANDROID_VERSIONS.join(", ")}.  If 3.6.1 is newly released, we may not have had a chance to release a version tested against it yet.  Proceed with caution.  You can suppress this warning with with -Dorg.gradle.android.cache-fix.ignoreVersionCheck=true./
+        }.size() == 1
```
Comment (runningcode) [inline src/test/groovy/org/gradle/android/PluginApplicationTest.groovy:44]: was this previously printing the output more than once?

**Review by runningcode — COMMENTED** [review]
looking good! just some questions. quick turnaround!

File: src/test/groovy/org/gradle/android/WorkaroundTest.groovy (hunk 6 above)
Comment (ghale) [reply src/test/groovy/org/gradle/android/WorkaroundTest.groovy:9]: There is a test in here for JDK8 and its associated versions and a test for JDK11 and its associated versions.

File: src/test/groovy/org/gradle/android/PluginApplicationTest.groovy (hunk 7 above)
Comment (ghale) [reply src/test/groovy/org/gradle/android/PluginApplicationTest.groovy:44]: Yep, once for the app module, once for the library module.

File: src/test/groovy/org/gradle/android/MultiVersionTest.groovy (hunk 5 above)
Comment (ghale) [reply src/test/groovy/org/gradle/android/MultiVersionTest.groovy:4]: Good catch.  I clearly got distracted...

File: src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy (hunk 3 above)
Comment (ghale) [reply src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy:528]: ¯\_(ツ)_/¯

**Review by runningcode — APPROVED** [review]
Looking good!

**Pull request merged**